### ISSUE TITLE
change the imgsz to 1280, training becomes extremely slow after 20 epochs

### ISSUE THREAD

**Issue by crescentxxx** [ultralytics/ultralytics#4276]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

When I first use imgsz=640 to train yolov8x on my own dataset, it seemed pretty fast and normal, although the training in the latter epochs alse became a little slow. Then I change imgsz to 1280, after training for about 20 epochs, it becomes very slow(less than  15minutes/epoch in the first several epochs, but more than one hour in the >20 epochs).
I have already searched the issues, including [this one](https://github.com/ultralytics/ultralytics/issues/3869), but does not seem helpful :(

in the early epochs:
![image](https://github.com/ultralytics/ultralytics/assets/49297123/8959e7be-c4ef-4e3c-a933-d91d1c33bd6d)

in the epoch26:
![fc9630b293d2e2df2883daa1618f003](https://github.com/ultralytics/ultralytics/assets/49297123/f018b4d5-0f20-45de-9121-16907a552d18)

## GPU
> GPU 0: NVIDIA GeForce RTX 3090 (UUID: GPU-d2d05a27-3918-5ab8-7275-8c05d997e126)
GPU 1: NVIDIA GeForce RTX 3090 (UUID: GPU-9947d96e-b6c9-e80f-5440-ae5a9032852d)
GPU 2: NVIDIA GeForce RTX 4090 (UUID: GPU-b1d264b9-8a0d-7473-1d88-22df5df95db7)
GPU 3: NVIDIA GeForce RTX 4090 (UUID: GPU-285459de-9454-68e7-c8fd-255999c829b0)
GPU 4: NVIDIA GeForce RTX 4090 (UUID: GPU-5f607a9b-28b3-e4e9-fe2f-9099a1636349)
GPU 5: NVIDIA GeForce RTX 3090 (UUID: GPU-a7213b2f-4c01-f97a-9141-3a61c4cf54f7)
GPU 6: NVIDIA GeForce RTX 3090 (UUID: GPU-311e2267-c7b4-583e-fc8f-9342243a3015)
GPU 7: NVIDIA GeForce RTX 4090 (UUID: GPU-3e22fb54-ab88-3295-92fd-febd589e3d04)

## Environment
Here is my environment:
> Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main    defaults
_openmp_mutex             5.1                       1_gnu    defaults
absl-py                   1.4.0                    pypi_0    pypi
aiohttp                   3.8.5                    pypi_0    pypi
aiohttp-retry             2.8.3                    pypi_0    pypi
aiosignal                 1.3.1                    pypi_0    pypi
amqp                      5.1.1                    pypi_0    pypi
antlr4-python3-runtime    4.9.3                    pypi_0    pypi
appdirs                   1.4.4                    pypi_0    pypi
async-timeout             4.0.2                    pypi_0    pypi
asyncssh                  2.13.2                   pypi_0    pypi
atpublic                  4.0                      pypi_0    pypi
attrs                     23.1.0                   pypi_0    pypi
automat                   22.10.0                  pypi_0    pypi
billiard                  4.1.0                    pypi_0    pypi
ca-certificates           2023.05.30           h06a4308_0    defaults
cachetools                5.3.1                    pypi_0    pypi
celery                    5.3.1                    pypi_0    pypi
certifi                   2023.7.22                pypi_0    pypi
cffi                      1.15.1                   pypi_0    pypi
charset-normalizer        3.2.0                    pypi_0    pypi
clearml                   1.12.1                   pypi_0    pypi
click                     8.1.6                    pypi_0    pypi
click-didyoumean          0.3.0                    pypi_0    pypi
click-plugins             1.1.1                    pypi_0    pypi
click-repl                0.3.0                    pypi_0    pypi
cmake                     3.27.0                   pypi_0    pypi
colorama                  0.4.6                    pypi_0    pypi
comet                     3.1.0                    pypi_0    pypi
configobj                 5.0.8                    pypi_0    pypi
constantly                15.1.0                   pypi_0    pypi
contourpy                 1.1.0                    pypi_0    pypi
cryptography              41.0.3                   pypi_0    pypi
cycler                    0.11.0                   pypi_0    pypi
dictdiffer                0.9.0                    pypi_0    pypi
diskcache                 5.6.1                    pypi_0    pypi
distro                    1.8.0                    pypi_0    pypi
dpath                     2.1.6                    pypi_0    pypi
dulwich                   0.21.5                   pypi_0    pypi
dvc                       3.12.0                   pypi_0    pypi
dvc-data                  2.11.0                   pypi_0    pypi
dvc-http                  2.30.2                   pypi_0    pypi
dvc-objects               0.24.1                   pypi_0    pypi
dvc-render                0.5.3                    pypi_0    pypi
dvc-studio-client         0.11.0                   pypi_0    pypi
dvc-task                  0.3.0                    pypi_0    pypi
dvclive                   2.13.0                   pypi_0    pypi
filelock                  3.12.2                   pypi_0    pypi
flatten-dict              0.4.2                    pypi_0    pypi
flufl-lock                7.1.1                    pypi_0    pypi
fonttools                 4.42.0                   pypi_0    pypi
frozenlist                1.4.0                    pypi_0    pypi
fsspec                    2023.6.0                 pypi_0    pypi
funcy                     2.0                      pypi_0    pypi
furl                      2.1.3                    pypi_0    pypi
gitdb                     4.0.10                   pypi_0    pypi
gitpython                 3.1.32                   pypi_0    pypi
google-auth               2.22.0                   pypi_0    pypi
google-auth-oauthlib      1.0.0                    pypi_0    pypi
grandalf                  0.8                      pypi_0    pypi
grpcio                    1.56.2                   pypi_0    pypi
hydra-core                1.3.2                    pypi_0    pypi
hyperlink                 21.0.0                   pypi_0    pypi
idna                      3.4                      pypi_0    pypi
importlib-metadata        6.8.0                    pypi_0    pypi
importlib-resources       6.0.0                    pypi_0    pypi
incremental               22.10.0                  pypi_0    pypi
iterative-telemetry       0.0.8                    pypi_0    pypi
jinja2                    3.1.2                    pypi_0    pypi
jsonschema                4.18.6                   pypi_0    pypi
jsonschema-specifications 2023.7.1                 pypi_0    pypi
kiwisolver                1.4.4                    pypi_0    pypi
kombu                     5.3.1                    pypi_0    pypi
ld_impl_linux-64          2.38                 h1181459_1    defaults
libffi                    3.4.4                h6a678d5_0    defaults
libgcc-ng                 11.2.0               h1234567_1    defaults
libgomp                   11.2.0               h1234567_1    defaults
libstdcxx-ng              11.2.0               h1234567_1    defaults
lit                       16.0.6                   pypi_0    pypi
lxml                      4.9.3                    pypi_0    pypi
markdown                  3.4.4                    pypi_0    pypi
markdown-it-py            3.0.0                    pypi_0    pypi
markupsafe                2.1.3                    pypi_0    pypi
matplotlib                3.7.2                    pypi_0    pypi
mdurl                     0.1.2                    pypi_0    pypi
multidict                 6.0.4                    pypi_0    pypi
nanotime                  0.5.2                    pypi_0    pypi
ncurses                   6.4                  h6a678d5_0    defaults
networkx                  3.1                      pypi_0    pypi
numpy                     1.25.2                   pypi_0    pypi
nvidia-cublas-cu11        11.10.3.66               pypi_0    pypi
nvidia-cublas-cu12        12.2.4.5                 pypi_0    pypi
nvidia-cuda-cupti-cu11    11.7.101                 pypi_0    pypi
nvidia-cuda-nvrtc-cu11    11.7.99                  pypi_0    pypi
nvidia-cuda-runtime-cu11  11.7.99                  pypi_0    pypi
nvidia-cuda-runtime-cu12  12.2.128                 pypi_0    pypi
nvidia-cudnn-cu11         8.5.0.96                 pypi_0    pypi
nvidia-cudnn-cu12         8.9.2.26                 pypi_0    pypi
nvidia-cufft-cu11         10.9.0.58                pypi_0    pypi
nvidia-curand-cu11        10.2.10.91               pypi_0    pypi
nvidia-cusolver-cu11      11.4.0.1                 pypi_0    pypi
nvidia-cusparse-cu11      11.7.4.91                pypi_0    pypi
nvidia-nccl-cu11          2.14.3                   pypi_0    pypi
nvidia-nvtx-cu11          11.7.91                  pypi_0    pypi
oauthlib                  3.2.2                    pypi_0    pypi
omegaconf                 2.3.0                    pypi_0    pypi
opencv-python             4.8.0.74                 pypi_0    pypi
openssl                   3.0.9                h7f8727e_0    defaults
orderedmultidict          1.0.1                    pypi_0    pypi
orjson                    3.9.2                    pypi_0    pypi
packaging                 23.1                     pypi_0    pypi
pafy                      0.5.5                    pypi_0    pypi
pandas                    2.0.3                    pypi_0    pypi
pathlib2                  2.3.7.post1              pypi_0    pypi
pathspec                  0.11.2                   pypi_0    pypi
pillow                    10.0.0                   pypi_0    pypi
pip                       23.2.1           py39h06a4308_0    defaults
platformdirs              3.10.0                   pypi_0    pypi
prompt-toolkit            3.0.39                   pypi_0    pypi
protobuf                  4.23.4                   pypi_0    pypi
psutil                    5.9.5                    pypi_0    pypi
py-cpuinfo                9.0.0                    pypi_0    pypi
pyasn1                    0.5.0                    pypi_0    pypi
pyasn1-modules            0.3.0                    pypi_0    pypi
pycparser                 2.21                     pypi_0    pypi
pydot                     1.4.2                    pypi_0    pypi
pygit2                    1.12.2                   pypi_0    pypi
pygments                  2.15.1                   pypi_0    pypi
pygtrie                   2.5.0                    pypi_0    pypi
pyjwt                     2.4.0                    pypi_0    pypi
pyparsing                 3.0.9                    pypi_0    pypi
python                    3.9.17               h955ad1f_0    defaults
python-dateutil           2.8.2                    pypi_0    pypi
pytz                      2023.3                   pypi_0    pypi
pyyaml                    6.0.1                    pypi_0    pypi
readline                  8.2                  h5eee18b_0    defaults
referencing               0.30.1                   pypi_0    pypi
requests                  2.31.0                   pypi_0    pypi
requests-oauthlib         1.3.1                    pypi_0    pypi
rich                      13.5.2                   pypi_0    pypi
rpds-py                   0.9.2                    pypi_0    pypi
rsa                       4.9                      pypi_0    pypi
ruamel-yaml               0.17.32                  pypi_0    pypi
ruamel-yaml-clib          0.2.7                    pypi_0    pypi
scipy                     1.11.1                   pypi_0    pypi
scmrepo                   1.1.0                    pypi_0    pypi
seaborn                   0.12.2                   pypi_0    pypi
setuptools                68.0.0           py39h06a4308_0    defaults
shortuuid                 1.0.11                   pypi_0    pypi
shtab                     1.6.4                    pypi_0    pypi
six                       1.16.0                   pypi_0    pypi
smmap                     5.0.0                    pypi_0    pypi
sqlite                    3.41.2               h5eee18b_0    defaults
sqltrie                   0.7.0                    pypi_0    pypi
tabulate                  0.9.0                    pypi_0    pypi
tensorboard               2.13.0                   pypi_0    pypi
tensorboard-data-server   0.7.1                    pypi_0    pypi
tensorrt-bindings         8.6.1                    pypi_0    pypi
tensorrt-libs             8.6.1                    pypi_0    pypi
tk                        8.6.12               h1ccaba5_0    defaults
tomlkit                   0.12.1                   pypi_0    pypi
torch                     2.0.1                    pypi_0    pypi
torchvision               0.15.2                   pypi_0    pypi
tqdm                      4.65.0                   pypi_0    pypi
triton                    2.0.0                    pypi_0    pypi
twisted                   22.10.0                  pypi_0    pypi
typing-extensions         4.7.1                    pypi_0    pypi
tzdata                    2023.3                   pypi_0    pypi
ultralytics               8.0.147                   dev_0    <develop>
urllib3                   1.26.16                  pypi_0    pypi
vine                      5.0.0                    pypi_0    pypi
voluptuous                0.13.1                   pypi_0    pypi
wcwidth                   0.2.6                    pypi_0    pypi
werkzeug                  2.3.6                    pypi_0    pypi
wheel                     0.38.4           py39h06a4308_0    defaults
xz                        5.4.2                h5eee18b_0    defaults
yarl                      1.9.2                    pypi_0    pypi
youtube-dl                2020.12.2                pypi_0    pypi
zc-lockfile               3.0.post1                pypi_0    pypi
zipp                      3.16.2                   pypi_0    pypi
zlib                      1.2.13               h5eee18b_0    defaults
zope-interface            6.0                      pypi_0    pypi

### training script
```python
from ultralytics import YOLO
from ultralytics import settings

model=YOLO('yolov8x.pt')
model.train(data='MyData.yaml', imgsz=1280)
```

### configuration yaml
> Train settings -------------------------------------------------------------------------------------------------------
model:  # (str, optional) path to model file, i.e. yolov8n.pt, yolov8n.yaml
data:  # (str, optional) path to data file, i.e. coco128.yaml
epochs: 100  # (int) number of epochs to train for
patience: 50  # (int) epochs to wait for no observable improvement for early stopping of training
batch: 64  # (int) number of images per batch (-1 for AutoBatch)
imgsz: 1280  # (int | list) input images size as int for train and val modes, or list[w,h] for predict and export modes
save: True  # (bool) save train checkpoints and predict results
save_period: 2 # (int) Save checkpoint every x epochs (disabled if < 1)
cache: False  # (bool) True/ram, disk or False. Use cache for data loading
device: [0,1,2,3,4,5,6,7] # (int | str | list, optional) device to run on, i.e. cuda device=0 or device=0,1,2,3 or device=cpu
workers: 8  # (int) number of worker threads for data loading (per RANK if DDP)
project:  # (str, optional) project name

**Comment by Kayzwer** [comment]
@crescentxxx it is actually normal, bigger input size (imgsz) = more time to finish feedforward.

**Comment by crescentxxx** [comment]
@Kayzwer But the first several epochs are pretty fast, about 15minutes per epoch. Why it gets so slow in the subsequent epochs instead of at the begining?

**Comment by glenn-jocher** [comment]
@Kayzwer it's natural for the training time per epoch to increase as the training progresses. This is due to the behavior of adaptive optimization methods like Adam and SGD with momentum, which are common in deep learning software. 

At the beginning of training, these optimizers make large updates to the model parameters, which can result in significant changes to the model's predictions. However, as training progresses and the model begins to converge, the updates become smaller and more fine-grained, requiring more computation.

Another reason could be due to batch normalization layers. These layers compute running statistics during training, which can also lead to longer training times in later epochs.

So, the more the model learns, the longer it takes for each epoch to train due to the finer adjustments being made. It's a sign that your model is improving its understanding of your data. Keep training.

**Comment by crescentxxx** [comment]
@glenn-jocher Thanks for your answer!

**Comment by rohansaw** [comment]
@glenn-jocher could you please elaborate, why the training time would increase due to larger updates?

In my understanding, while the value of the updates might be smaller or larger, this should not influence the computational time to do the actual update. The same applies to batch normalization - in the classical implementation of batch norm, the computational time of the statistics should be roughly constant over the epochs.

**Comment by glenn-jocher** [comment]
@rohansaw sure! The observation that training time increases might not be directly due to the magnitude of updates or the computation in batch normalization (BN) itself being slower over time. 

Instead, the increasing time per epoch might stem from factors like:
- **Data Loading and Augmentation**: As the model fits better to the training data, the benefit from increasingly complex augmentations might result in slightly prolonged data preprocessing times.
- **Dynamic Learning Rate Adjustments**: Alterations in learning rates can affect training dynamics and might subtly influence training time per epoch if adaptive computation is engaged.
- **IO and Memory Bound Processes**: Depending on the system and training setup, disk I/O for checkpoint saving, data loading, or memory swapping (if applicable) can exhibit variability in time usage over epochs.

The essence here is that while the fundamental operations per forward and backward pass remain computationally constant, the ecosystem around these passes (including data handling and system-level operations) might introduce variability in epoch durations.

I hope this clarifies things a bit! 🚀
